README.md
# LogDNA CLI

<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-14-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

The LogDNA CLI allows you to sign up for a new account and tail your logs right from the command line.

## Deprecation Notice: LogDNA CLI deprecates supporting CentOS 6 and Debian 8 as of v2.

## Getting Started

Note that prebuilt images are of [v2.0.0](https://github.com/logdna/logdna-cli/tree/2.0.0) (without natural language time), not of [master](https://github.com/logdna/logdna-cli/tree/master).  Please bear with us as we convert to an automated release pipeline for this repo.

### macOS

Download the [LogDNA CLI installer for Mac](http://repo.logdna.com/mac/logdna-cli.pkg).  

Alternatively install via [brew cask](https://caskroom.github.io/):
```
brew cask update
brew cask install logdna-cli
```

### Windows

To install on Windows, using [chocolatey](https://chocolatey.org):

```
choco install logdna
```

### Linux

For Ubuntu/Debian:

```bash
sudo wget -qO /tmp/logdna-cli.deb http://repo.logdna.com/linux/logdna-cli.deb && sudo dpkg -i /tmp/logdna-cli.deb
```

For RHEL/CentOS:

```bash
sudo wget -qO /tmp/logdna-cli.rpm http://repo.logdna.com/linux/logdna-cli.rpm && sudo rpm -ivh /tmp/logdna-cli.rpm
```

### From Source

Follow these quick instructions to run the LogDNA CLI from source:

```
git clone https://github.com/logdna/logdna-cli.git
cd logdna-cli
npm install --production

# help
node index.js --help

# login using one of commands below
node index.js login
node index.js ssologin
```

## Usage

| Command | Description |
| - | - |
| `logdna register [email]` | Register a new LogDNA account |
| `logdna ssologin` | Log in to a LogDNA via single sign-on |
| `logdna login [email]` | Log in to LogDNA |
| `logdna tail [options] [query]` | Live tail with optional filtering. Options include `-h`, `-a`, `-l`, `-t` to filter by hosts, apps, levels or tags respectively. Run `logdna tail --help` to learn more. |
| `logdna switch` | If your login has access to more than one account, this command allows you to switch between them |
| `logdna search [options] [query]` | Basic search with optional filtering. Run `logdna search --help` for options. |
| `logdna info` | Show current logged in user info |
| `logdna update` | Update the CLI to the latest version |

### Examples

```sh
# Register
$ logdna register user@example.com

# Login
$ logdna login user@example.com

# Tail
$ logdna tail '("timed out" OR "connection refused") -request'
$ logdna tail -a access.log 500
$ logdna tail -l error,warn

# Search
$ logdna search "logdna cli" -a logdna.log -t tag1,tag2 -n 300
# Search via UNIX timestamps
$ logdna search "logdna" --from 1541100040931 --to 1541102940000
# Search via Natural Language (Chrono)
# Ensure to enclose natural langauge strings in ""
# Must include date reference (i.e., `May 26`, `yesterday`, or the like)
$ logdna search "logdna" --from "yesterday at 3pm" --to "today at 3:05pm"
$ logdna search "logdna" --timeframe "May 26 at 3pm to 3:05pm"
# Search with dev logs
$ logdna search "logdna" --timeframe "May 28 at 1pm to 3:05pm" -d

# Other
$ logdna info
$ logdna update

# Switch organization
$ logdna switch

> 1: ACME (active)
> 2: Strickland Propane
> Choose account [1-2]:
```

## Developer Logs
Logs for troubleshooting during development are included. These logs will pipe to console and optionally to [LogDNA](https://www.logdna.com/).

Switch dev logs on via the `-d` or `--dev` flag.

In order to pipe the developer logs to LogDNA, you should set the account API Key using `LOGDNA_API_KEY` environment variable.  Otherwise, dev logs will only been seen in the console.

Currently the dev log coverage is limited to the code regarding `--timeframe`, `--from` and `--to` in [index.js](./index.js).


## Contributing

Contributions are always welcome. See the [contributing guide](./CONTRIBUTING.md) to learn how you can help.

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/yields"><img src="https://avatars.githubusercontent.com/u/1661587?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Amir Abushareb</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=yields" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=yields" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/braxtonj"><img src="https://avatars.githubusercontent.com/u/7331755?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Braxton</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=braxtonj" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=braxtonj" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/meffij"><img src="https://avatars.githubusercontent.com/u/8787479?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Calvin Cochran</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=meffij" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=meffij" title="Documentation">📖</a></td>
    <td align="center"><a href="https://ernesti.me/"><img src="https://avatars.githubusercontent.com/u/20255948?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Ernest Iliiasov</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=ernestii" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=ernestii" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/jakedipity"><img src="https://avatars.githubusercontent.com/u/29671917?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Jacob Hull</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=jakedipity" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=jakedipity" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/stuckj"><img src="https://avatars.githubusercontent.com/u/2205578?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Jonathan Stucklen</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=stuckj" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=stuckj" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/leeliu"><img src="https://avatars.githubusercontent.com/u/1399797?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Lee Liu</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=leeliu" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=leeliu" title="Documentation">📖</a> <a href="https://github.com/logdna/logdna-cli/commits?author=leeliu" title="Tests">⚠️</a></td>
  </tr>
  <tr>
    <td align="center"><a href="https://github.com/mikehu"><img src="https://avatars.githubusercontent.com/u/981800?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Mike Hu</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=mikehu" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/respectus"><img src="https://avatars.githubusercontent.com/u/1046364?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Muaz Siddiqui</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=respectus" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=respectus" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/beefcheeks"><img src="https://avatars.githubusercontent.com/u/4133158?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Ryan</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=beefcheeks" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=beefcheeks" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/smusali"><img src="https://avatars.githubusercontent.com/u/34287490?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Samir Musali</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=smusali" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=smusali" title="Documentation">📖</a> <a href="https://github.com/logdna/logdna-cli/commits?author=smusali" title="Tests">⚠️</a></td>
    <td align="center"><a href="http://stevenedouard.com/"><img src="https://avatars.githubusercontent.com/u/3053263?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Steven Edouard</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=sedouard" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=sedouard" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/ydshah2"><img src="https://avatars.githubusercontent.com/u/11240269?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Yash Shah</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=ydshah2" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=ydshah2" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/vilyapilya"><img src="https://avatars.githubusercontent.com/u/17367511?v=4?s=100" width="100px;" alt=""/><br /><sub><b>vilyapilya</b></sub></a><br /><a href="https://github.com/logdna/logdna-cli/commits?author=vilyapilya" title="Code">💻</a> <a href="https://github.com/logdna/logdna-cli/commits?author=vilyapilya" title="Documentation">📖</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed
- Add periodic websocket pings when connected to livetail. This adds support for aggresive proxies that disconnect idle connections
- Natural Language Processing on timestamps using [Chrono](https://github.com/wanasit/chrono).  Introduction of --timeframe option for search.  Addition of developer logging via LogDNA for troubleshooting.
- Changed CONTRIBUTING.md to reflect "fork-and-PR" process

## [v2.0.0] - 2020-05-21
### Fixed
- Fix `Cannot read property 'statusCode' of undefined` when updating [#40](https://github.com/logdna/logdna-cli/pull/40)

### Changed
- Write information output to stderr for tail and search [#43](https://github.com/logdna/logdna-cli/pull/43)
- Reorganize and modify package info [#45](https://github.com/logdna/logdna-cli/pull/45)

## [v1.4.1] - 2019-08-14
### Changed
- Don't show error if the configuration file does not exist [#38](https://github.com/logdna/logdna-cli/pull/38)

## [v1.4.0] - 2019-08-13
### Changed
- Don't require super user [#36](https://github.com/logdna/logdna-cli/pull/36)

### Added
- Add Usage Information to Readme [#35](https://github.com/logdna/logdna-cli/pull/35)

### Fixed
- Update packages and fix authentication [#33](https://github.com/logdna/logdna-cli/pull/33)

## [v1.3.1] - 2019-02-14
### Fixed
- Fix SSO Login [#25](https://github.com/logdna/logdna-cli/pull/25)
- Fix NPM Vulnerabilities [#24](https://github.com/logdna/logdna-cli/pull/24)
- Fix issue with search where results with only 1 result aren't properly displayed [#23](https://github.com/logdna/logdna-cli/pull/23)

### Changed
- Use `request` instead of `got` [#22](https://github.com/logdna/logdna-cli/pull/22)

## [v1.3.0] - 2018-11-09
### Fixed
- Fix the issues and update the CLI [#21](https://github.com/logdna/logdna-cli/pull/21)
- Fix incorrect query parameter to the search API [#16](https://github.com/logdna/logdna-cli/pull/16)

## [v1.2.2] - 2017-10-11
### Added
- Add SSO support to CLI

## [v1.2.1] - 2017-09-25
### Added
- Add --json option [#12](https://github.com/logdna/logdna-cli/pull/12)

## [v1.2.0] - 2017-05-25
### Added
- `switch` option
- ANSI color for levels

### Changed
- `logdna search` now uses `Export API`

## [v1.1.1] - 2017-02-03
### Fixed
- Fix tailing

### Changed
- ANSI color support for `tail`/`search`
- Specify or omit email on command line support for both `login` and `register`

### Added
- Add Kubernetes and Docker instructions
- Add `-h` help option for `tail`/`search`

## [v1.1.0] - 2016-08-24
### Fixed
- Fix various authentication issues

### Added
- Add instructions for Syslog, Heroku, API, Node.js
- Add `brew cask` install option for Mac [#5](https://github.com/logdna/logdna-cli/pull/5)

## [v1.0.9] - 2016-03-29
### Added
- Add Windows and Mac support for CLI
- Add instructions for Windows Hosts by doing `logdna install windows` [#1](https://github.com/logdna/logdna-cli/pull/1)


[v2.0.0]: https://github.com/answerbook/logdna-workers/compare/v1.4.1...v2.0.0
[v1.4.1]: https://github.com/answerbook/logdna-workers/compare/v1.4.0...v1.4.1
[v1.4.0]: https://github.com/answerbook/logdna-workers/compare/v1.3.1...v1.4.0
[v1.3.1]: https://github.com/answerbook/logdna-workers/compare/v1.3.0...v1.3.1
[v1.3.0]: https://github.com/answerbook/logdna-workers/compare/v1.2.2...v1.3.0
[v1.2.2]: https://github.com/answerbook/logdna-workers/compare/v1.2.1...v1.2.2
[v1.2.1]: https://github.com/answerbook/logdna-workers/compare/v1.2.0...v1.2.1
[v1.2.0]: https://github.com/answerbook/logdna-workers/compare/v1.1.1...v1.2.0
[v1.1.1]: https://github.com/answerbook/logdna-workers/compare/v1.0.9...v1.1.1
[v1.1.0]: https://github.com/answerbook/logdna-workers/compare/v1.0.9...v1.1.0
[v1.0.9]: https://github.com/answerbook/logdna-workers/releases/tag/v1.0.9


lib/utils.js
// External Modules
const crypto = require('crypto');
const os = require('os');
const properties = require('properties');
const qs = require('querystring');
const request = require('request');
const semver = require('semver');
const spawn = require('child_process').spawn;

// Internal Modules
const pkg = require('../package.json');

// Constants
const CLI_USER_AGENT = 'logdna-cli/' + pkg.version;
const ERROR_LEVEL_TEST = /[Ee]rr(?:or)?|ERR(?:OR)?|[Cc]rit(?:ical)?|CRIT(?:ICAL)?|[Ff]atal|FATAL|[Ss]evere|SEVERE|EMERG(?:ENCY)?|[Ee]merg(?:ency)?/;
const WARN_LEVEL_TEST = /[Ww]arn(?:ing)?|WARN(?:ING)?/;
const FORCE_UPDATE_TIMEOUT = 30000;
const DEFAULT_UPDATE_TIMEOUT = 2500;

const saveConfig = function(config, callback) {
    properties.stringify(config, {
        path: config.DEFAULT_CONF_FILE
    }, function(err) {
        if (err) return callback('Error while saving to: ' + (config.DEFAULT_CONF_FILE) + ': ' + err);
        return callback(null, true);
    });
};

const uiDisp = function(msg, config, level) {
    try {
        if (level === 'error') {
            console.error(msg || '');
        } else {
            console.log(msg || '');
        }
    } catch (e) {
        if (config) {
            saveConfig(config, (error, result) => {
                if (error) console.error(error);
                return process.exit(0);
            });
        }
    }
};

function generateHmac(payload, secret) {
    let msg = qs.stringify(payload);
    return crypto.createHmac('sha256', secret.toString()).update(msg).digest('hex');
};

const authParams = function(config) {
    if (!config.token) {
        uiDisp('Please login first. Type \'logdna login\' or \'logdna --help\' for more info.');
        return process.exit();
    }

    let hmacParams = {
        email: config.email
        , id: config.account
        , ts: Date.now()
    };

    hmacParams.hmac = generateHmac(hmacParams, config.token);
    return hmacParams;
};

function apiCall(config, endpoint, method, params, callback) {
    if (typeof params === 'function') {
        callback = params;
        params = null;
    }

    let opts = {
        method: method
        , headers: {
            'content-type': 'application/json'
            , 'user-agent': CLI_USER_AGENT
        }
    };

    params = params || {};
    if (params.auth === false) { // Don't use auth (i.e. for login, register)
        delete params.auth;
    } else if (typeof params.auth === 'string') { // Use HTTP auth
        if (params.auth) {
            const [user, pass] = params.auth.split(':');
            opts.auth = { user, pass };
        }
        delete params.auth;
    } else { // Use HMAC token stored in config
        const hmacParams = authParams(config);
        opts.qs = hmacParams;
    }

    opts.url = (config.LOGDNA_APISSL ? 'https://' : 'http://') + config.LOGDNA_APIHOST + '/' + endpoint;
    if (method === 'get') {
        opts.qs = Object.assign(opts.qs || {}, params);
    } else {
        opts.body = JSON.stringify(params);
    }

    request(opts, (error, response, body) => {
        try {
            body = JSON.parse(body);
        } catch (err) {
            // Keep the body as is
        }

        if (error || body.error) {
            if (response && response.statusCode === 403) {
                return callback('Access token invalid. If you created or changed your password recently, ' +
                                'please run \'logdna login\' again. Type \'logdna --help\' for more info.');
            }

            return callback(`Error: ${error || body.error}`);
        }

        callback(null, body);
    });
};

function levelColor(config, level) {
    if (ERROR_LEVEL_TEST.test(level)) return '\x1b[38;5;255;48;5;203m';
    if (WARN_LEVEL_TEST.test(level)) return '\x1b[38;5;255;48;5;208m';
    return '\x1b[38;5;247;48;5;239m';
};

const apiGet = function(config, endpoint, params, callback) {
    apiCall(config, endpoint, 'get', params, callback);
};

const apiPost = function(config, endpoint, params, callback) {
    apiCall(config, endpoint, 'post', params, callback);
};

const renderLine = function(config, line, params) {
    // ignore bad payloads
    if (!line._host) return;

    let t = new Date(line._ts);
    params = params || {};

    if (params.json) return uiDisp(JSON.stringify(line));

    if (config.SUPPORTS_COLORS) {
        uiDisp('\x1b[38;5;240m' + t.toString().substring(4, 11) + t.toString().substring(16, 24) +
            ' \x1b[38;5;166m' + line._host +
            ' \x1b[38;5;74m' + line._app +
            ' ' + (line.level ? levelColor(config, line.level) + ' ' + line.level + ' \x1b[0m ' : '') +
            '\x1b[38;5;246m' + (line.message || line._line) +
            '\x1b[0m', config);
    } else {
        uiDisp(t.toString().substring(4, 11) + t.toString().substring(16, 24) +
            ' ' + line._host +
            ' ' + line._app +
            ' ' + (line.level ? '[' + line.level + '] ' : '') +
            (line.message || line._line), config);
    };
};

const performUpgrade = function(config, force, callback) {
    if (typeof force === 'function') {
        callback = force;
        force = false;
    }

    const updateCheckIntervalPassed = Date.now() - (config.updatecheck || 0) > config.UPDATE_CHECK_INTERVAL;
    const shouldUpdate = force || updateCheckIntervalPassed;
    if (!shouldUpdate) {
        return callback(); // no check necessary, run rest of program
    } else {
        if (os.platform() === 'darwin') {
            config.VERSION_CHECK_URL = config.VERSION_CHECK_URL.replace('PLATFORM', 'mac');
            config.UPDATE_CHECK_URL = config.UPDATE_CHECK_URL.replace('PLATFORM', 'mac');
        } else if (os.platform() === 'linux') {
            config.VERSION_CHECK_URL = config.VERSION_CHECK_URL.replace('PLATFORM', 'linux');
            config.UPDATE_CHECK_URL = config.UPDATE_CHECK_URL.replace('PLATFORM', 'linux');
        } else if (os.platform() === 'win32') {
            // config.VERSION_CHECK_URL = config.VERSION_CHECK_URL.replace("PLATFORM", "windows");
            // config.UPDATE_CHECK_URL = config.UPDATE_CHECK_URL.replace("PLATFORM", "windows");
            return callback(); // ignore for now until this works
        }

        if (force) { uiDisp('Checking for updates...'); }
        const timeout = force ? FORCE_UPDATE_TIMEOUT : DEFAULT_UPDATE_TIMEOUT;
        request.get(config.VERSION_CHECK_URL, { timeout }, (error, response, body) => {
            if (error) {
                return uiDisp(`Error${response ? ' [' + response.statusCode + ']' : ''}: ${error}`);
            }

            if (body) { body = body.replace(/\r/g, '').replace(/\n/g, ''); }

            if (!semver.valid(body)) {
                config.updatecheck = Date.now() - config.UPDATE_CHECK_INTERVAL + 86400000;
                return saveConfig(config, function(error, success) {
                    if (error) return uiDisp(error) && callback(error);
                    return callback();
                });
            };

            config.updatecheck = Date.now();
            saveConfig(config, function(error, success) {
                if (error) return uiDisp(error);
            });

            if (semver.gt(body, pkg.version)) {
                // update needed
                uiDisp('Performing upgrade from ' + pkg.version + ' to ' + body + '...');
                let shell = spawn('/bin/bash', ['-c'
                    , 'if [[ ! -z $(which curl) ]]; then curl -so /tmp/logdna.gz ' + config.UPDATE_CHECK_URL + '; elif [[ ! -z $(which wget) ]]; then wget -qO /tmp/logdna.gz ' + config.UPDATE_CHECK_URL + '; fi; gunzip -f /tmp/logdna.gz; cp -f /tmp/logdna /usr/local/logdna/bin/logdna; chmod 777 /usr/local/logdna/bin/logdna 2> /dev/null; echo -n "Successfully upgraded logdna-cli to "; /usr/local/bin/logdna -v'
                ], {
                    stdio: 'inherit'
                });

                shell.on('close', function() {
                    if (!force && process.argv[process.argv.length - 1].toLowerCase() !== 'update') uiDisp('Please run your command again');
                });

                return;
            } else {
                return callback(); // no update necessary, run rest of program
            }
        });
    }
};

// Convenience function to create standardized developer log object
const devLogPrim = function(opts) {
    const {
        message = '' // Message to be displayed in LogDNA LiveTail. Additional fields/objects will be stored, searchable and actionable
        , who = 'dev' // Who is sending this log. Often user but here it's dev
        , what = '' // High level "what" is happening
        , why = '' // Why the log is occuring
        , where = 'local' // Where in your system this is happening
        , when = new Date().toISOString() // When this occured in your system. Can be different then the time of the log itself
        , additional = {} // Additional fields to include in the log. An object is required
    } = opts;

    return {
        message
        , who
        , what
        , why
        , where
        , when
        , ...additional
    };
}

let module_exports = {
    apiGet
    , apiPost
    , authParams
    , uiDisp
    , performUpgrade
    , renderLine
    , saveConfig
    , devLogPrim
};

module.exports = module_exports;
